File: components/fxkit/code/FXMultiFoldingView.m
/**
This file is part of the Volta project.
Copyright (C) 2007-2013 Kai Berk Oezer
https://robo.fish/wiki/index.php?title=Volta
https://github.com/robo-fish/Volta

Volta is free software. You can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#import "FXMultiFoldingView.h"
#import "FXClipView.h"
#import "FXViewUtils.h"


#pragma mark - FXMultiFoldingSubview -


@interface FXMultiFoldingSubview : FXClipView
{
@private
  NSButton* mTriangleButton;
  NSTextField* mTitleField;
  NSSize mMinSize;
  FXClipView* mContainerView;
  id<NSObject> __weak mTarget;
}
@property SEL action;
@property (weak) id<NSObject> target;
@property (readonly) BOOL isFolded;
@property (copy) NSString* title;

- (id) initWithTitle:(NSString*)title;
- (void) setContentView:(NSView*)view;
- (void) handleDisclosureAction:(id)sender;
@end

static const CGFloat skFoldingSubviewHeaderHeight = 18.0;


@implementation FXMultiFoldingSubview

@synthesize target = mTarget;

- (id) initWithTitle:(NSString*)title
{
  static const NSRect dummyFrame = {0,0,200,200};
  if ( (self = [super initWithFrame:dummyFrame]) != nil )
  {
    mMinSize = NSZeroSize;

    NSView* view = [[NSView alloc] initWithFrame:dummyFrame];

    mTitleField = [[NSTextField alloc] initWithFrame:dummyFrame];
    mTitleField.stringValue = title;
    mTitleField.bordered = NO;
    mTitleField.backgroundColor = [NSColor colorWithDeviceRed:0.4 green:0.42 blue:0.4 alpha:1];
    mTitleField.drawsBackground = YES;
    mTitleField.textColor = [NSColor colorWithDeviceWhite:0.9 alpha:1.0];
    mTitleField.font = [NSFont fontWithName:@"Lucida Grande" size:12.0];
    mTitleField.selectable = NO;
    mTitleField.editable = NO;
    mTitleField.focusRingType = NSFocusRingTypeNone;
    mTitleField.alignment = NSTextAlignmentCenter;

    mContainerView = [[FXClipView alloc] initWithFrame:dummyFrame];

    view.subviews = @[mTitleField, mContainerView];
    FXRelease(mTitleField)
    FXRelease(mContainerView)

    NSDictionary* views = @{ @"container" : mContainerView, @"title" : mTitleField };
    NSDictionary* metrics = @{ @"titleHeight" : @(skFoldingSubviewHeaderHeight) };
    [FXViewUtils layoutInView:view
                visualFormats:@[@"H:|[title]|", @"H:|[container]|", @"V:|[title(titleHeight)][container(>=32)]|"]
                  metricsInfo:metrics
                    viewsInfo:views];

    self.documentView = view;
    self.minDocumentViewHeight = skFoldingSubviewHeaderHeight;
    FXRelease(view)
  }
  return self;
}

- (void) dealloc
{
  FXDeallocSuper
}

- (void) setTitle:(NSString*)title
{
  [mTitleField setStringValue:title];
}

- (NSString*) title
{
  return [mTitleField stringValue];
}

- (BOOL) isFolded
{
  return ([mTriangleButton state] == NSOffState);
}

- (void) setContentView:(NSView*)view
{
  mContainerView.documentView = view;
  NSDictionary* views = NSDictionaryOfVariableBindings(view);
  [mContainerView removeConstraints:[mContainerView constraints]];
  [FXViewUtils layoutInView:mContainerView visualFormats:@[@"H:|[view]|", @"V:|[view]|"] metricsInfo:nil viewsInfo:views];
  self.minDocumentViewHeight = skFoldingSubviewHeaderHeight + view.fittingSize.height;
}

- (void) handleDisclosureAction:(id)sender
{
  NSAssert(sender == mTriangleButton, @"Action sent from unknown control.");
  if ( (self.target != nil) && (self.action != 0) )
  {
  #pragma clang diagnostic push
  #pragma clang diagnostic ignored "-Warc-performSelector-leaks"
    [self.target performSelector:self.action];
  #pragma clang diagnostic pop
  }
}

- (void) encodeRestorableStateWithCoder:(NSCoder*)coder
{
  [super encodeRestorableStateWithCoder:coder];
  [mContainerView encodeRestorableStateWithCoder:coder];
}

- (void) restoreStateWithCoder:(NSCoder*)coder
{
  [super restoreStateWithCoder:coder];
  [mContainerView restoreStateWithCoder:coder];
}

@end


#pragma mark - FXMultiFoldingView -


@interface FXMultiFoldingView () <NSSplitViewDelegate>
@end


@implementation FXMultiFoldingView
{
@private
  NSSplitView* mSplitter;
  NSMutableArray* mFoldingSubviews;
}

- (id)initWithFrame:(NSRect)frame
{
  if ( (self = [super initWithFrame:frame]) != nil )
  {
    mSplitter = [[NSSplitView alloc] initWithFrame:frame];
    mSplitter.dividerStyle = NSSplitViewDividerStylePaneSplitter;
    mSplitter.delegate = self;
    mSplitter.vertical = NO;
    [super addSubview:mSplitter];
    FXRelease(mSplitter)

    [FXViewUtils layoutInView:self visualFormats:@[@"H:|[mSplitter]|", @"V:|[mSplitter]|"] metricsInfo:nil viewsInfo:NSDictionaryOfVariableBindings(mSplitter)];

    mFoldingSubviews = [NSMutableArray new];
  }
  return self;
}


- (void)dealloc
{
  FXRelease(mFoldingSubviews)
  FXDeallocSuper
}


#pragma mark Public


- (void) addSubview:(NSView*)view withTitle:(NSString*)title
{
  FXMultiFoldingSubview* foldingSubview = [self newFoldingSubviewWithTitle:title];
  foldingSubview.contentView = view;
  [mSplitter addSubview:foldingSubview];
  [FXViewUtils layoutInView:mSplitter
              visualFormats:@[@"V:[foldingSubview(>=minHeight)]"]
                metricsInfo:@{@"minHeight":@(foldingSubview.minDocumentViewHeight)}
                  viewsInfo:NSDictionaryOfVariableBindings(foldingSubview)];
  [mFoldingSubviews addObject:foldingSubview];
  FXRelease(foldingSubview)
}


- (void) addSubview:(NSView*)view
{
  [self addSubview:view withTitle:@""];
}


#pragma mark NSSplitViewDelegate methods


- (BOOL) splitView:(NSSplitView*)splitView canCollapseSubview:(NSView*)subview
{
  return YES;
}


- (BOOL) splitView:(NSSplitView *)splitView shouldCollapseSubview:(NSView*)subview forDoubleClickOnDividerAtIndex:(NSInteger)dividerIndex
{
  return YES;
}


FXIssue(247)
#if 0
- (BOOL) splitView:(NSSplitView *)splitView shouldAdjustSizeOfSubview:(NSView*)targetSubview
{
  CGFloat const totalDividerThickness = MAX([mFoldingSubviews count] - 1, 0) * [mSplitter dividerThickness];
  CGFloat const totalHeight = self.frame.size.height;
  CGFloat const subviewHeight = targetSubview.frame.size.height;
  BOOL const allOtherSubviewsAreCollapsed = ( ( totalHeight - subviewHeight - totalDividerThickness) < 1 );
  return allOtherSubviewsAreCollapsed;
}
#endif


#pragma mark NSResponder overrides


NSString* FXResume_SplitterPositions = @"FXResume_SplitterPositions";

- (void) encodeRestorableStateWithCoder:(NSCoder*)coder
{
  [super encodeRestorableStateWithCoder:coder];
  NSArray* splitterSubviews = [mSplitter subviews];
  __block NSMutableArray* subviewHeights = [[NSMutableArray alloc] initWithCapacity:[splitterSubviews count]];
  [splitterSubviews enumerateObjectsUsingBlock:^(id subview, NSUInteger index, BOOL *stop) {
    FXMultiFoldingSubview* foldingSubview = subview;
    [foldingSubview encodeRestorableStateWithCoder:coder];
    NSNumber* heightNumber = @([foldingSubview frame].size.height);
    [subviewHeights addObject:heightNumber];
  }];
  [coder encodeObject:subviewHeights forKey:FXResume_SplitterPositions];
  FXRelease(subviewHeights)
  subviewHeights = nil;
}


- (void) restoreStateWithCoder:(NSCoder*)coder
{
  [super restoreStateWithCoder:coder];
  NSArray* subviewHeights = [coder decodeObjectForKey:FXResume_SplitterPositions];
  for ( FXMultiFoldingSubview* subview in [mSplitter subviews] )
  {
    [subview restoreStateWithCoder:coder];
  }
  [mSplitter adjustSubviews];
  __block CGFloat currentPos = 0.0;
  [subviewHeights enumerateObjectsUsingBlock:^(id heightNumber, NSUInteger index, BOOL *stop) {
    if ( index < ([[mSplitter subviews] count] - 1) )
    {
      currentPos += [(NSNumber*)heightNumber doubleValue] + (index * [mSplitter dividerThickness]);
      [mSplitter setPosition:currentPos ofDividerAtIndex:index];
    }
  }];
  [self setNeedsDisplay:YES];
}


#pragma mark Private


- (void) handleFoldingAction:(id)sender
{

}


- (FXMultiFoldingSubview*) newFoldingSubviewWithTitle:(NSString*)title
{
  FXMultiFoldingSubview* foldingBox = [[FXMultiFoldingSubview alloc] initWithTitle:title];
  foldingBox.action = @selector(handleFoldingAction:);
  foldingBox.target = self;
  return foldingBox;
}


@end

File: components/fxkit/code/FXMultiFoldingView.swift
/**
This file is part of the Volta project.
Copyright (C) 2018 Kai Berk Oezer
https://robo.fish/wiki/index.php?title=Volta
https://github.com/robo-fish/Volta

Volta is free software. You can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
import Cocoa

public class FXMultiFoldingView : NSView
{
  private var _splitter : NSSplitView
  private var _foldingSubviews = [FXMultiFoldingSubview]()

  public override init(frame frameRect: NSRect)
  {
    _splitter = NSSplitView(frame:frameRect)
    _splitter.dividerStyle = .paneSplitter
    _splitter.isVertical = false
    super.init(frame: frameRect)
    _splitter.delegate = self
    addSubview(_splitter)
    FXViewUtils.layout(in: self, visualFormats:["H:|[splitter]|", "V:|[splitter]|"], metricsInfo:nil, viewsInfo:["splitter":_splitter])
  }

  public convenience required init?(coder decoder: NSCoder)
  {
    self.init(frame: .zero)
  }

  /// Appends the given view to the bottom of the receiver.
  @objc public func addSubview(_ view : NSView, withTitle title : String)
  {
    let foldingSubview = _newFoldingSubview(title:title)
    _splitter.addSubview(foldingSubview)
    foldingSubview.contentView = view
    FXViewUtils.layout(in:_splitter,
              visualFormats:["V:[foldingSubview(>=minHeight)]"],
                metricsInfo:["minHeight":NSNumber(value:Double(foldingSubview.minDocumentViewHeight))],
                  viewsInfo:["foldingSubview":foldingSubview])
    _foldingSubviews.append(foldingSubview)
  }

  @objc func handleFoldingAction(_ sender : Any)
  {
  }

  private func _newFoldingSubview(title : String) -> FXMultiFoldingSubview
  {
    let foldingBox = FXMultiFoldingSubview(title: title)
    foldingBox.action = #selector(handleFoldingAction(_:))
    foldingBox.target = self
    return foldingBox
  }
}

let FXResume_SplitterPositions = "FXResume_SplitterPositions"

extension FXMultiFoldingView
{
  public override func restoreState(with coder: NSCoder)
  {
    super.restoreState(with: coder)
    guard let subviewHeights = coder.decodeObject(forKey: FXResume_SplitterPositions) as? [NSNumber] else { return }
    _splitter.subviews.forEach({ $0.restoreState(with:coder) })
    _splitter.adjustSubviews()
    var currentPos = CGFloat(0)
    for (index, item) in subviewHeights.enumerated()
    {
      if index < (self._splitter.subviews.count - 1)
      {
        currentPos += CGFloat(item.doubleValue) + CGFloat(index)*_splitter.dividerThickness
        _splitter.setPosition(currentPos, ofDividerAt: index)
      }
    }
    self.needsDisplay = true
  }

  public override func encodeRestorableState(with coder: NSCoder)
  {
    super.encodeRestorableState(with: coder)
    var subviewHeights = [NSNumber]()
    _splitter.subviews.forEach {
      $0.encodeRestorableState(with: coder)
      subviewHeights.append(NSNumber(value:Double($0.frame.size.height)))
    }
    coder.encode(subviewHeights, forKey:FXResume_SplitterPositions)
  }
}

extension FXMultiFoldingView : NSSplitViewDelegate
{
  public func splitView(_ splitView: NSSplitView, canCollapseSubview subview: NSView) -> Bool
  {
    return true
  }

  public func splitView(_ splitView: NSSplitView, shouldCollapseSubview subview: NSView, forDoubleClickOnDividerAt dividerIndex: Int) -> Bool
  {
    return true
  }

//  func splitView(_ splitView: NSSplitView, shouldAdjustSizeOfSubview subview: NSView) -> Bool
//  {
//    let totalDividerThickness = max(CGFloat(_foldingSubviews.count - 1), 0) * _splitter.dividerThickness;
//    let totalHeight = self.frame.size.height
//    let subviewHeight = subview.frame.size.height
//    let allOtherSubviewsAreCollapsed = ( ( totalHeight - subviewHeight - totalDividerThickness) < 1 )
//    return allOtherSubviewsAreCollapsed
//  }
}

let skFoldingSubviewHeaderHeight = CGFloat(18.0)

class FXMultiFoldingSubview : FXClipView
{
  private var _triangleButton : NSButton?
  private var _titleField : NSTextField
  private var _minSize = CGSize.zero
  private var _containerView : FXClipView

  var action : Selector?
  weak var target : AnyObject?
  var isFolded : Bool {
    return _triangleButton?.state == .off
  }

  var title : String
  {
    set { _titleField.stringValue = newValue }
    get { return _titleField.stringValue }
  }

  init(title : String)
  {
    let dummyFrame = CGRect(x:0, y:0, width:200, height:200)

    _titleField = NSTextField(frame:dummyFrame)
    _titleField.stringValue = title
    _titleField.isBordered = false
    _titleField.backgroundColor = NSColor(deviceRed:0.4, green: 0.42, blue: 0.4, alpha: 1.0)
    _titleField.drawsBackground = true
    _titleField.textColor = NSColor(deviceWhite:0.9, alpha:1.0)
    _titleField.font = NSFont(name:"Lucida Grande", size:12.0)
    _titleField.isSelectable = false
    _titleField.isEditable = false
    _titleField.focusRingType = .none
    _titleField.alignment = .center

    _containerView = FXClipView(frame:dummyFrame)

    let view = NSView(frame:dummyFrame)
    view.subviews = [_titleField, _containerView]

    FXViewUtils.layout(in:view,
      visualFormats:["H:|[title]|", "H:|[container]|", "V:|[title(titleHeight)][container(>=32)]|"],
      metricsInfo:["titleHeight": NSNumber(value:Double(skFoldingSubviewHeaderHeight))],
      viewsInfo:["container":_containerView, "title":_titleField])

    super.init(frame: dummyFrame, flipped:false)
    self.title = title
    self.documentView = view
    self.minDocumentViewHeight = skFoldingSubviewHeaderHeight
  }

  required convenience init?(coder decoder: NSCoder)
  {
    let title = (decoder.decodeObject(forKey: "Title") as? String) ?? ""
    self.init(title:title)
  }

  var contentView : NSView?
  {
    set {
      _containerView.documentView = newValue
      _containerView.removeConstraints(_containerView.constraints)
      if let view = newValue
      {
        FXViewUtils.layout(in:_containerView, visualFormats:["H:|[view]|", "V:|[view]|"], metricsInfo:nil, viewsInfo:["view":view])
        self.minDocumentViewHeight = skFoldingSubviewHeaderHeight + view.fittingSize.height
      }
    }
    get {
      return _containerView.documentView
    }
  }

  func handleDisclosureAction(sender : AnyObject)
  {
    assert(sender === _triangleButton, "Action sent from unknown control.")
    if let target_ = self.target, let action_ = self.action
    {
      target_.performSelector(onMainThread: action_, with: nil, waitUntilDone: false)
    }
  }

  override func encodeRestorableState(with coder: NSCoder)
  {
    _containerView.encodeRestorableState(with: coder)
    super.encodeRestorableState(with: coder)
  }

  override func restoreState(with coder: NSCoder)
  {
    super.restoreState(with: coder)
    _containerView.restoreState(with: coder)
  }
}
